src/input/shared.rs

```diff
@@ -136,3 +137,22 @@ pub fn float_as_int<'a>(input: &'a impl Input<'a>, float: f64) -> ValResult<'a,
         Err(ValError::new(ErrorTypeDefaults::IntParsingSize, input))
     }
 }
+
+pub fn decimal_as_int<'a>(py: Python, input: &'a impl Input<'a>, decimal: &PyAny) -> ValResult<'a, EitherInt<'a>> {
+    if !decimal.call_method0(intern!(py, "is_finite"))?.extract::<bool>()?
+        || decimal.call_method0(intern!(py, "is_nan"))?.extract::<bool>()?
```

Comment:
`is_finite` is sufficient to check for NaN too

```suggestion
    if !decimal.call_method0(intern!(py, "is_finite"))?.extract::<bool>()?
```

```python
>>> Decimal('nan').is_finite()
False
```